OCPBUGS-11146: update cloud provider flags on vSphere for storage options

elmiko opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**

* modify the render pipeline to include the storage operator object
* update the cloud provider flag to check the storage options when using vSphere
* only use external cloud provider flag on vSphere when CSI storage is enabled

**- How to verify it**

there are 3 main conditions to verify with this change:

1. starting with a 4.12 vSphere cluster using in-tree storage, upgrade to 4.13 should preserve the `--cloud-provider=vsphere` flag on the kubelet while using `--cloud-provider=external` on the kube-apiserver and kube-controller-manager.
2. starting with a 4.12 vSphere cluster using out-of-tree storage (CSI), upgrade to 4.13 should upgrade all cloud provider flags to `--cloud-provider=external`.
3. installing new clusters on 4.13 should use all cloud provider flags as `--cloud-provider=external`.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Update external cloud provider check for vSphere to observe the CSI migration status and properly configure the cluster for in-tree storage and out-of-tree cloud providers.

elmiko commented:
i am still working on this but wanted to start sharing for others to see what we are proposing.

elmiko commented on pkg/controller/template/render.go:
```diff
@@ -391,9 +396,38 @@ func skipMissing(key string) (interface{}, error) {
+		if err != nil {
+			return external, nil
+		}
+		isexternal := external && ((storageConfig == nil) || (storageConfig != nil && storageConfig.Spec.VSphereStorageDriver == operatorv1.CSIWithMigrationDriver))
```
i'm not sure if this is the proper logic for cases where `storageConfig == nil`, will we ever have a condition where that storage api is non-existent?

cc @jsafrane @dobsonj

elmiko replied:
thanks @jsafrane !

elmiko commented on pkg/controller/template/template_controller.go:
```diff
@@ -117,16 +123,24 @@ func New(
+		//AddFunc:    ctrl.addFeature,
+		//UpdateFunc: ctrl.updateFeature,
+		//DeleteFunc: ctrl.deleteFeature,
+	})
```
i'm still working on these

dobsonj replied:
Won't this cause problems if for example `addFeature` is called and provided a `Storage` object which `addFeature` casts to a `FeatureGate`?
https://github.com/openshift/machine-config-operator/blob/f08e554770af2c879ea32a68c3d66c068ee0da83/pkg/controller/template/template_controller.go#L219-L223

elmiko commented:
i think this PR is ready for review, i am going to begin manually testing it.

elmiko commented:
switched to the release-4.13 branch and rebased this PR to match.

also added a commit to handle the `OPENSHIFT_DO_VSPHERE_MIGRATION` environment variable.

i am still wrestling with the unit tests.

dobsonj commented on pkg/controller/template/template_controller.go:
```diff
@@ -117,16 +123,24 @@ func New(
 		DeleteFunc: ctrl.deleteFeature,
 	})
 
+	storageInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addFeature,
+		UpdateFunc: ctrl.updateFeature,
+		DeleteFunc: ctrl.deleteFeature,
+	})
```
Won't this cause problems if for example `addFeature` is called and provided a `Storage` object which `addFeature` casts to a `FeatureGate`?
https://github.com/openshift/machine-config-operator/blob/f08e554770af2c879ea32a68c3d66c068ee0da83/pkg/controller/template/template_controller.go#L219-L223

elmiko replied:
yeah, i think i was misunderstanding those functions when i uncommented the originals. i'll work up functions for those.

dobsonj commented on pkg/controller/template/render.go:
```diff
@@ -708,3 +743,11 @@ func isOpenShiftManagedDefaultLB(cfg RenderConfig) bool {
 }
+
+func isVSphereCSIMigrationEnabled(cfg RenderConfig) interface{} {
+	if cfg.StorageConfig.Spec.VSphereStorageDriver == operatorv1.CSIWithMigrationDriver {
+		return "enabled"
+	}
```
Do you need to handle the case where StorageConfig is nil, similar to `isCloudProviderExternal`?
https://github.com/openshift/machine-config-operator/pull/3655/files#diff-f02b383ddd2e30a342d61ceabe5d3c30742a0f169680cf09a449c3b53ffd6babR420

elmiko replied:
yes, good spot

elmiko commented:
updated with latest comments

* add functions for storage handlers
* check for storage nil object in render func

elmiko commented:
updated to fix up the unit tests and add tests for the new logic.

elmiko commented:
i am  manually testing this on a 4.13 cluster with the other storage related changes and it looks like it is doing what we want on a default install. i can see that the kubelet has `--cloud-provider=vsphere` and the kcm has `--cloud-provider=external --external-cloud-volume-plugin=vsphere`.

gnufied commented on pkg/controller/template/template_controller_test.go:
```diff
@@ -475,7 +487,7 @@ func TestUpdateMachineConfig(t *testing.T) {
 	}
 
-	expmcs, err := getMachineConfigsForControllerConfig(templateDir, cc, []byte(`{"dummy": "dummy"}`), feat)
+	expmcs, err := getMachineConfigsForControllerConfig(templateDir, cc, []byte(`{"dummy": "dummy"}`), feat, nil)
```
We should get a test here that storage object is populated and passed correctly.

elmiko replied:
sounds like a good upgrade, i'll take a look

elmiko commented:
i was wrong about the expected default condition, and it looks like we might have hit a condition where the Storage object is nil going into the logic. i have added the Storage lister to the function that caches the informer, i am going to test manually.

elmiko commented:
updated to add rbac for the storage object to the machine-config-controller

JoelSpeed commented:
LGTM to me apart from the requested additional storage test

elmiko commented on pkg/controller/template/template_controller.go:
```diff
@@ -341,6 +355,37 @@ func (ctrl *Controller) deleteMachineConfig(obj interface{}) {
+	oldStorage := old.(*operatorv1.Storage)
+	glog.V(4).Infof("Updating Storage %s", oldStorage.Name)
+	ctrl.enqueueController()
+}
```
we are seeing an issue where cluster installations are not happening properly and the master nodes are coming up with a different configuration than expected. it appears like the Storage object is `nil` when the master machineconfig templates are created, and i wonder if this function needs a change to ensure they get updated?

JoelSpeed replied:
The issue I believe is with the bootstrap, see my comment about decoding and the fact we don't have the infra object

elmiko commented:
updated with the latest changes from #3668 and squashed commits to make it cleaner

elmiko commented:
i've added a bootstrap e2e test for the storage object, but i'm really uncertain about the way this test is supposed to work. i'm guessing it will need some adjustment.

elmiko commented:
fixing an issue with the test manifest loader

elmiko commented:
updated to add Storages to the clientset for the test environment.

elmiko commented:
adding missing create function for Storage cr

elmiko commented:
add missing initializer for Operator lister in test clientset

elmiko commented:
trying out an experiment moving the storage test, it seems like there is a containerruntimeconfig in the test environment (even though it should be removed). i'm wondering if it is leftover from the previous test case.

elmiko commented:
i'm not sure what generated files should be expected by the test here. i had thought there would be some sort of updated kubelet unit file.

JoelSpeed commented on pkg/controller/template/render.go:
```diff
@@ -391,9 +397,38 @@ func skipMissing(key string) (interface{}, error) {
+// https://github.com/kubernetes/kubernetes/pull/116342
+// https://issues.redhat.com/browse/STOR-1265
+func isCloudProviderExternal(platformStatus *configv1.PlatformStatus, featureGate *configv1.FeatureGate, storageConfig *operatorv1.Storage) (bool, error) {
+	if platformStatus == nil {
+		return false, fmt.Errorf("platformStatus is required")
+	}
```
At the moment, the render for bootstrap exits here because platformStatus is nil, we need to fix that for the render to work correctly

elmiko replied:
i don't think this is correct, we check for nil status before passing in the `cloudProvider`  in both places we use the function.

elmiko commented on pkg/controller/bootstrap/bootstrap.go:
```diff
@@ -76,6 +77,7 @@ func (b *Bootstrap) Run(destDir string) error {
 	var cconfig *mcfgv1.ControllerConfig
 	var featureGate *apicfgv1.FeatureGate
 	var nodeConfig *apicfgv1.Node
+	var storageConfig *apioperatorsv1.Storage
```
ack, i'll add the operatorv1 group, the last entry for `apicfgv1` should be the configv1

elmiko commented:
added some fixes to the bootstrap tests and also a missed addition of v1 operators to the json decoder.

i am having trouble when i take out the check for is the storage error is not found, there needs to be more work on the unit tests to make them work. but, we can at least check that this patch works with the others.

dobsonj commented:
/lgtm
This is working for me and the latest changes look good.
/hold for QE to finish pre-merge testing

dobsonj commented:
/retitle OCPBUGS-11146: update cloud provider flags on vSphere for storage options
(just to make the bot happy with a 4.13 bug in the title)

yuqi-zhang commented on pkg/controller/bootstrap/testdata/bootstrap/machineconfigcontroller-controllerconfig.yaml:
```diff
@@ -32,6 +32,9 @@ spec:
       controlPlaneTopology: HighlyAvailable
       platformStatus:
         type: None
+  dns:
```
Just curious, was this added for a test? I can't seem to match what this is being used for

elmiko replied:
some of the vsphere templates need the dns field to render properly

yuqi-zhang commented on pkg/controller/template/template_controller.go:
```diff
@@ -480,7 +524,13 @@ func (ctrl *Controller) syncControllerConfig(key string) error {
 	}
-	mcs, err := getMachineConfigsForControllerConfig(ctrl.templatesDir, cfg, pullSecretRaw, fg)
+	storageConfig, err := ctrl.storageConfLister.Get(ctrlcommon.ClusterFeatureInstanceName)
+	if err != nil && !errors.IsNotFound(err) {
```
I may have missed the conclusion of this discussion. If the object is not yet created during install, or we have some weird flake, we would bypass this and call the next function but with a (presumably) nil storageConfig. Would that cause any problems? Or is that fine/we won't run into those situations?

elmiko replied:
we are defaulting to the CSI migration is the storage config is nil, i tried to remove this `IsNotFound` check but it broke the unit tests mightily, and i need more time to fix those. we decided that the effect was limited enough that we could work on this specific piece as part of a separate bug [OCPBUGS-11833](https://issues.redhat.com/browse/OCPBUGS-11833).

yuqi-zhang commented on pkg/controller/template/template_controller.go:
```diff
@@ -507,6 +557,8 @@ func getMachineConfigsForControllerConfig(templatesDir string, config *mcfgv1.Co
 		ControllerConfigSpec: &config.Spec,
 		PullSecret:           string(buf.Bytes()),
 		FeatureGate:          featureGate,
+
```
nit: weird formatting here with extra line and offset indentation it seems?

yuqi-zhang commented on templates/master/01-master-kubelet/_base/files/kubelet.yaml:
```diff
@@ -28,7 +28,6 @@ contents:
       APIPriorityAndFairness: true
       RotateKubeletServerCertificate: true
       DownwardAPIHugePages: true
-      OpenShiftPodSecurityAdmission: true
```
Question: is this due to the api update?

elmiko replied:
yes, it got moved in the openshift/api repo between 4.13 and 4.14. but now that i think about it, not sure why we had to move it in 4.13 here, iirc the tests were broken until we changed the template.

yuqi-zhang reviewed (commented):
Generally seems fine to me from an MCO logic perspective, some questions below:

Also, for non-vsphere platforms, we should see no difference in-cluster right?

elmiko commented:
> Also, for non-vsphere platforms, we should see no difference in-cluster right?

correct

elmiko commented:
fixed formatting issue in template_controller.go

dobsonj commented:
/lgtm

/assign @yuqi-zhang
for approval

knobunc commented:
/label backport-risk-assessed
/label cherry-pick-approved

yuqi-zhang reviewed (approved):
It sounds like this was tested extensively, so I think I am fine giving the approval. The general workflow lgtm although I haven't personally tested on vsphere. So long as we don't break any use cases (which CI should cover) I think this is fine

dobsonj commented:
/unhold
